subprojects/plugins/src/integTest/groovy/org/gradle/groovy/compile/GroovyCompileToolchainIntegrationTest.groovy

```diff
@@ -0,0 +1,211 @@
+
+        then:
+        executedAndNotSkipped(":compileGroovy")
+        outputContains("Compiling with JDK Java compiler API")
+        targetJdk.javaVersion == JavaVersion.forClass(groovyClassFile("Foo.class").bytes)
+        groovyTarget(targetJdk.javaVersion) == JavaVersion.forClass(groovyClassFile("Bar.class").bytes)
+
+        where:
+        what             | when                         | withTool | withJavaHome | withExecutable | withJavaExtension | target
+        "current JVM"    | "when nothing is configured" | null     | null         | null           | null              | "current"
+        "java extension" | "when configured"            | null     | null         | null           | "other"           | "other"
+        "assigned tool"  | "when configured"            | "other"  | null         | null           | null              | "other"
+        "assigned tool"  | "over java extension"        | "other"  | null         | null           | "current"         | "other"
+    }
+
+    def "up-to-date depends on the toolchain"() {
+        def currentJdk = Jvm.current()
+        def otherJdk = AvailableJavaHomes.getDifferentVersion()
+
+        buildFile << """
+            compileGroovy {
+                javaLauncher = javaToolchains.launcherFor {
+                    languageVersion = JavaLanguageVersion.of(
+                        providers.gradleProperty("changed").isPresent()
+                            ? ${otherJdk.javaVersion.majorVersion}
+                            : ${currentJdk.javaVersion.majorVersion}
+                    )
+                }
+            }
+        """
+
+        when:
+        withInstallations(currentJdk, otherJdk).run(":compileGroovy")
+        then:
+        executedAndNotSkipped(":compileGroovy")
+
+        when:
+        withInstallations(currentJdk, otherJdk).run(":compileGroovy")
+        then:
+        skipped(":compileGroovy")
+
+        when:
+        withInstallations(currentJdk, otherJdk).run(":compileGroovy", "-Pchanged", "--info")
+        then:
+        executedAndNotSkipped(":compileGroovy")
+        outputContains("Value of input property 'groovyCompilerJvmVersion' has changed for task ':compileGroovy'")
+        outputContains("Value of input property 'javaLauncher.metadata.languageVersion' has changed for task ':compileGroovy'")
+
+        when:
+        withInstallations(currentJdk, otherJdk).run(":compileGroovy", "-Pchanged")
+        then:
+        skipped(":compileGroovy")
+    }
+
+    def 'source and target compatibility override toolchain (source #source, target #target)'() {
+        def jdk11 = AvailableJavaHomes.getJdk(JavaVersion.VERSION_11)
+
+        buildFile << """
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(11)
+                }
+            }
+
+            compileGroovy {
+                if ("$source" != 'none')
+                    sourceCompatibility = JavaVersion.toVersion($source)
```

Comment:
```suggestion
                ${source != 'none'
                    ? "sourceCompatibility = JavaVersion.toVersion($source)"
                    : ''}
```
🤔  I think it is better to move the if outside of the build script. Same for the other ones.

Reply:
Done